### PR TITLE
Unify Loggers on Interfaces and Deprecate Implemenations

### PR DESCRIPTION
**Pull request by anthony-murphy**
This change deprecates our internal TelemetryLogger implementations and unifies our exported and consumed surface area on our telemetry interfaces.

For the deprecated implementations the following replacement function should be used:

-   replace ChildLogger.create, new TelemetryNullLogger, and new BaseTelemetryNullLogger with createChildLogger
-   replace DebugLogger.create with createDebugLogger
-   replace new MultiSinkLogger with createMultiSinkLogger
-   replace TelemetryUTLogger with MockLogger

### CODE REVIEW

File: packages/utils/telemetry-utils/src/logger.ts [hunk 1]
```diff
@@ -276,20 +285,42 @@ export class TaggedLoggerAdapter implements ITelemetryBaseLogger {
 				default:
 					// If we encounter a tag we don't recognize
 					// then we must assume we should scrub.
-					newEvent[key] = "REDACTED (unknown tag)";
+					newEvent[key] = `REDACTED (unknown tag)`;
```
Comment (anthony-murphy) [inline packages/utils/telemetry-utils/src/logger.ts:288]:
```suggestion
					newEvent[key] = "REDACTED (unknown tag)';
```

File: packages/loader/container-utils/src/test/errorTypeLoggingTest.spec.ts [hunk 2]
```diff
@@ -97,9 +97,12 @@ describe("Check if the errorType field matches after sending/receiving via Conta
 		});
 	});
 
-	describe("Send errors using a ChildLogger", () => {
+	describe("Send errors using a createChildLogger", () => {
```
Comment (alexvy86) [inline packages/loader/container-utils/src/test/errorTypeLoggingTest.spec.ts:100]: Nit
```suggestion
	describe("Send errors using a logger from createChildLogger", () => {
```

**Comment by markfields** [comment]
What's the impetus for this change?

**Comment by anthony-murphy** [comment]
> What's the impetus for this change?

@markfields  seeing incorrect type usage in api, like in dev tools. i also want to make some changes that are harder to make with so many leaked implementations. this will centralize on our interfaces rather than needing to support both interfaces and concrete types. Lastly, i think i can delete a bunch of code if we stop leaking our implmentations.

File: azure/packages/test/scenario-runner/src/logger.ts [hunk 3]
```diff
@@ -120,20 +121,24 @@ export async function getLogger(
...
 		baseLogger.transformEvents(transformEvents);
 	}
-	return ChildLogger.create(baseLogger, config.namespace, {
-		all: {
-			runId: config.runId,
-			scenarioName: config.scenarioName,
-			endpoint: config.endpoint,
-			region: config.region,
+	return createChildLogger({
+		logger: baseLogger,
```
Comment (markfields) [inline azure/packages/test/scenario-runner/src/logger.ts:133]: Subjective nit:  Maybe call the param property `baseLogger` instead of `logger`?

File: packages/drivers/local-driver/src/localDocumentService.ts [hunk 4]
```diff
@@ -96,6 +98,8 @@ export class LocalDocumentService implements IDocumentService {
 			ordererToken.jwt,
 			client,
 			this.localDeltaConnectionServer.webSocketServer,
+			undefined,
+			this.logger,
```
Comment (markfields) [inline packages/drivers/local-driver/src/localDocumentService.ts:102]: So this is a behavior change where we'll start collecting more logs whereas before they were going to the ether?  Nice to separate that in a change this big.

File: azure/packages/test/scenario-runner/src/logger.ts (hunk 3 above)
Comment (anthony-murphy) [reply azure/packages/test/scenario-runner/src/logger.ts:133]: it takes more bits, and i don't think it provides much more clarity, so i'd rather stick with the shorter name

File: packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts [hunk 5]
```diff
@@ -265,7 +265,7 @@ export class OdspDocumentServiceFactoryCore
 
 	protected async createDocumentServiceCore(
```
Comment (anthony-murphy) [inline packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts:266]: reverting this, as it is breaking technically

File: packages/drivers/odsp-driver/src/test/createNewUtilsTests.spec.ts [hunk 6]
```diff
@@ -93,7 +93,7 @@ describe("Create New Utils Tests", () => {
 				docId: hashedDocumentId,
 				resolvedUrl,
 			},
-			new TelemetryNullLogger(),
+			createChildLogger(),
```
Comment (markfields) [inline packages/drivers/odsp-driver/src/test/createNewUtilsTests.spec.ts:96]: It's a little weird to call `createChildLogger()` with no params.  If you think about it, of course it means the logs sent to it go nowhere, but I think many people may glance at it and not realize this is a no-op logger.

File: packages/utils/telemetry-utils/src/logger.ts [hunk 7]
```diff
@@ -561,6 +613,7 @@ export class PerformanceEvent {
 /**
  * Logger that is useful for UT
  * It can be used in places where logger instance is required, but events should be not send over.
+ * @deprecated - Use createChildLogger instead
  */
 export class TelemetryUTLogger implements ITelemetryLoggerExt {
```
Comment (alexvy86) [inline packages/utils/telemetry-utils/src/logger.ts:618]: Replacing this with `createChildLogger()` is a loss of behavior, right? The tests using it would stop reporting errors to console.

File: packages/drivers/odsp-driver/src/test/createNewUtilsTests.spec.ts (hunk 6 above)
Comment (anthony-murphy) [reply packages/drivers/odsp-driver/src/test/createNewUtilsTests.spec.ts:96]: it is also weird to have test code in a production package, since this is test, i'd rather the test do weird things, than have the production package do weird things.

File: packages/loader/container-loader/src/container.ts [hunk 8]
```diff
@@ -473,7 +473,7 @@ export class Container
 	private readonly codeLoader: ICodeDetailsLoader;
 	private readonly options: ILoaderOptions;
 	private readonly scope: FluidObject;
-	private readonly subLogger: TelemetryLogger;
+	private readonly subLogger: ITelemetryLoggerExt;
```
Comment (markfields) [inline packages/loader/container-loader/src/container.ts:476]: Oh, right, the Ext types.  I think earlier I saw `ITelemetryLogger` plain, should probably be `ITelemetryLoggerExt`, so people can use arrays (and flat objects in next major)

File: packages/drivers/odsp-driver/src/test/createNewUtilsTests.spec.ts (hunk 6 above)
Comment (alexvy86) [reply packages/drivers/odsp-driver/src/test/createNewUtilsTests.spec.ts:96]: I feel the same as Mark. Wouldn't necessarily block the change, but aleast we need to document the new functions, and especially this one to clarify what happens when called with no params.

File: packages/tools/replay-tool/src/replayMessages.ts [hunk 9]
```diff
@@ -168,7 +172,7 @@ class Document {
 	private documentSeqNumber = 0;
 	private from: number = -1;
 	private snapshotFileName: string = "";
-	private docLogger: TelemetryLogger;
+	private docLogger: ITelemetryLogger;
```
Comment (markfields) [inline packages/tools/replay-tool/src/replayMessages.ts:175]:
```suggestion
	private docLogger: ITelemetryLoggerExt;
```

File: packages/utils/telemetry-utils/src/logger.ts [hunk 10]
```diff
@@ -220,21 +230,22 @@ export abstract class TelemetryLogger implements ITelemetryLoggerExt {
...
+						if (eventLike[key] !== undefined) {
 							continue;
 						}
 						const getterOrValue = props[key];
 						// If this throws, hopefully it is handled elsewhere
 						const value =
 							typeof getterOrValue === "function" ? getterOrValue() : getterOrValue;
 						if (value !== undefined) {
-							newEvent[key] = value;
+							// @ts-expect-error what
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:241]: ?

File: packages/utils/telemetry-utils/src/logger.ts (hunk 7 above)
Comment (anthony-murphy) [reply packages/utils/telemetry-utils/src/logger.ts:618]: we hide console logs by default in tests, so i'm not sure how much value that provides, i really just want all the test loggers out of this production package. we could copy this to where it is used, rather than using child loggers, but not sure it is worth it since we try to avoid console logs in tests

File: packages/utils/telemetry-utils/src/logger.ts [hunk 11]
```diff
@@ -340,6 +376,7 @@ export class ChildLogger extends TelemetryLogger {
 		);
 	}
 
+	private readonly [childLoggerType] = childLoggerType;
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:379]: I'm assuming this is to allow ChildLogger coallescince across layers at different versions where `instanceof` doesn't work?  Maybe put in a different PR

File: packages/utils/telemetry-utils/src/logger.ts [hunk 12]
```diff
@@ -284,12 +295,37 @@ export class TaggedLoggerAdapter implements ITelemetryBaseLogger {
 	}
 }
 
+const childLoggerType = "__fluid_child_logger" as const;
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:298]: I wonder if we have convention on the formatting of this kind of tag.  Might be nice

File: packages/drivers/odsp-driver/src/test/createNewUtilsTests.spec.ts (hunk 6 above)
Comment (anthony-murphy) [reply packages/drivers/odsp-driver/src/test/createNewUtilsTests.spec.ts:96]: i also don't think people are very intentional about which test logger they use, looks more like they picked one, and the rest of the test just keep using it.

File: packages/utils/telemetry-utils/src/logger.ts [hunk 13]
```diff
@@ -284,12 +295,37 @@ export class TaggedLoggerAdapter implements ITelemetryBaseLogger {
...
+
+export function createChildLogger(props?: {
+	logger?: ITelemetryBaseLogger;
+	namespace?: string;
+	properties?: ITelemetryLoggerPropertyBags;
+}): ITelemetryLoggerExt {
+	if (props === undefined) {
+		return new TelemetryNullLogger();
+	}
+	return ChildLogger.create(props?.logger, props?.namespace, props?.properties);
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:308]:
```suggestion
	return ChildLogger.create(props.logger, props.namespace, props.properties);
```

Or, just remove the `if` check above and let there be one codepath since apparently `ChildLogger` can do it.  And remove `TelemetryNullLogger` altogether

File: packages/utils/telemetry-utils/src/logger.ts [hunk 14]
```diff
@@ -284,12 +295,37 @@ export class TaggedLoggerAdapter implements ITelemetryBaseLogger {
...
  * Creates sub-logger that appends properties to all events
+ * @deprecated - Use createChildLogger instead
  */
 export class ChildLogger extends TelemetryLogger {
+	private static is(logger?: ITelemetryBaseLogger): logger is ChildLogger {
+		if (logger instanceof ChildLogger) {
+			return true;
+		}
+		const maybe = logger as ChildLogger | undefined;
+		if (maybe?.[childLoggerType] === childLoggerType) {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:323]: super nit:  Just return this condition

File: packages/utils/telemetry-utils/src/logger.ts [hunk 15]
```diff
@@ -363,19 +400,34 @@ export class ChildLogger extends TelemetryLogger {
 	}
 }
 
+export function createMultiSinkLogger(props: {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:403]: Doc comments on exported functions are always nice

File: packages/loader/container-loader/src/container.ts (hunk 8 above)
Comment (anthony-murphy) [reply packages/loader/container-loader/src/container.ts:476]: i didn't worry about moving all internal references, and some land on public interfaces where it doesn't work. where i added new public exposures i used TelemetryUTLogger. i expect someone will do all the interface clean up in next soon.

File: packages/runtime/container-runtime/src/summary/runningSummarizer.ts [hunk 16]
```diff
@@ -200,7 +201,7 @@ export class RunningSummarizer implements IDisposable {
 		const maxAckWaitTime = Math.min(this.configuration.maxAckWaitTime, maxSummarizeAckWaitTime);
 
 		this.pendingAckTimer = new PromiseTimer(maxAckWaitTime, () => {
-			// Note: summarizeCount (from ChildLogger definition) may be 0,
+			// Note: summarizeCount (from createChildLogger definition) may be 0,
```
Comment (alexvy86) [inline packages/runtime/container-runtime/src/summary/runningSummarizer.ts:204]: Nit: the original comment doesn't make sense to me, is the update necessary?

File: packages/utils/telemetry-utils/src/logger.ts (hunk 7 above)
Comment (markfields) [reply packages/utils/telemetry-utils/src/logger.ts:618]: (I never liked this class - I think it's better to use `MockLogger` and be explicit about failing the test, rather than blindly throwing and hoping it breaks whatever you're asserting in the test)

File: packages/drivers/odsp-driver/src/odspUtils.ts [hunk 17]
```diff
@@ -14,7 +14,7 @@ import {
 } from "@fluidframework/driver-utils";
 import { assert, performance } from "@fluidframework/common-utils";
 import {
-	createChildLogger,
+	ChildLogger,
```
Comment (markfields) [inline packages/drivers/odsp-driver/src/odspUtils.ts]: Accidental revert?

File: packages/utils/telemetry-utils/src/logger.ts (hunk 7 above)
Comment (markfields) [reply packages/utils/telemetry-utils/src/logger.ts:618]: Ok but responding to the actual comment - `reportError` also throws.  So it's a pretty big change, I wouldn't do it here.

Comment (anthony-murphy) [reply packages/utils/telemetry-utils/src/logger.ts:618]: i missed that that logger threw, and didn't just  output it, i will revert these changes, as not throwing is a regression

File: packages/drivers/odsp-driver/src/odspUtils.ts (hunk 17 above)
Comment (anthony-murphy) [reply packages/drivers/odsp-driver/src/odspUtils.ts]: no. this publicly exposes TelemetryLogger via Childlogger.create, so it would be a break to change in main

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts [hunk 18]
```diff
@@ -59,8 +61,8 @@ export class LocalDocumentDeltaConnection extends DocumentDeltaConnection {
 		return deltaConnection;
 	}
 
-	constructor(socket: Socket, documentId: string) {
-		super(socket, documentId, new TelemetryNullLogger());
+	constructor(socket: Socket, documentId: string, logger?: ITelemetryBaseLogger) {
+		super(socket, documentId, createChildLogger({ logger }));
```
Comment (alexvy86) [inline packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:65]: Nit: I think this could actually be changing behavior slightly, not 100% sure. Inside `createChildLogger()`, even if the `logger` here is undefined, this will not qualify as undefined props but an empty object, right? If so, it will go through a slightly different path and end up returning a `ChildLogger` with a `TelemetryBaseNullLogger` as base instead of directly a `TelemetryNullLogger`.

File: packages/runtime/container-runtime/src/summary/runningSummarizer.ts (hunk 16 above)
Comment (anthony-murphy) [reply packages/runtime/container-runtime/src/summary/runningSummarizer.ts:204]:
```suggestion
			// Note: summarizeCount (from ChildLogger definition) may be 0,
```

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts (hunk 18 above)
Comment (anthony-murphy) [reply packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:65]: that is correct, but i think the null logger here was wrong, as it will lead to us loosing logs in tests. i think the change is minor. happy to put in a different review if you think it is necessary.

File: packages/utils/telemetry-utils/src/logger.ts [hunk 19]
```diff
@@ -284,10 +295,26 @@ export class TaggedLoggerAdapter implements ITelemetryBaseLogger {
 	}
 }
 
+/**
+ * Create a child logger based on the provided props object
+ * @param props - logger is the base logger the child will log to after it's processing, namespace will be prefixed to all event names, properties are default properties that will be applied events.
+ */
```
Comment (alexvy86) [inline packages/utils/telemetry-utils/src/logger.ts:301]:
```suggestion
/**
 * Create a child logger based on the provided props object
 * @param props - logger is the base logger the child will log to after it's processing, namespace will be prefixed to all event names, properties are default properties that will be applied events.
 *
 * @remarks
 * Passing in no props object (i.e. undefined) will return a logger that is effectively a no-op.
 */
```

File: packages/drivers/local-driver/src/localDocumentDeltaConnection.ts (hunk 18 above)
Comment (alexvy86) [reply packages/drivers/local-driver/src/localDocumentDeltaConnection.ts:65]: > happy to put in a different review if you think it is necessary.

Not really, if there was intention behind the change I don't have objections.

**Review by alexvy86 — APPROVED** [review]
One docs suggestion below, but LGTM.

File: packages/loader/container-loader/src/container.ts (hunk 8 above)
Comment (anthony-murphy) [reply packages/loader/container-loader/src/container.ts:476]: it also changes which package the interface comes from, which i didn't want to do everywhere.
